Set affinity mask via testbed file.

semihalf-zmuda-wojciech opened the pull request:
Implement setup_coremask_node() and call it during vRouter setup
to get affinity mask from the testbed file and put it into the
contrail-vrouter-dpdk.ini file.

Change-Id: Id68c7889b39889188088b35209682f46d9535ebd

semihalf-zmuda-wojciech committed:
Set affinity mask via testbed file.

Implement setup_coremask_node() and call it during vRouter setup
to get affinity mask from the testbed file and put it into the
contrail-vrouter-dpdk.ini file.

Change-Id: Id68c7889b39889188088b35209682f46d9535ebd

majkijin commented on fabfile/tasks/helpers.py:
```diff
@@ -806,6 +806,31 @@ def setup_hugepages_node(*args):
+        else:
+            return
+
+        if (coremask == ""):
```
Shouldn't we protect not only for the empty coremask but also for other "too small" like 2 cores? As far as I know we need at least 4 cores to run DPDK vRouter.

majkijin commented on fabfile/tasks/helpers.py:
```diff
@@ -806,6 +806,31 @@ def setup_hugepages_node(*args):
+            coremask = default_coremask
+
+        with settings(host_string=host_string):
+                sudo('sudo sed -i \'s/__COREMASK__/%s/\' %s' %(coremask, vrouter_file))
```
You don't need sudo twice here. Also, this breaks the contrail-vrouter-dpdk-init package (see comment in the packaging repo pull request).

semihalf-berestovskyy-andriy replied:
We can use 'sed s/taskset ${default_coremask}/taskset ${coremask}/', i.e. to always use the default mask in the contrail-vrouter-dpdk.ini file?

semihalf-zmuda-wojciech committed:
Post review fixes

semihalf-zmuda-wojciech committed:
Add support for listing CPUs.

We support two ways of specifying affinity mask:
- hexadecimal, eg. 0x3f
- list of CPUs, eg. 0,1,2,4-6

semihalf-zmuda-wojciech committed:
Post review fixes.

srajag commented on fabfile/tasks/helpers.py:
```diff
@@ -806,6 +806,38 @@ def setup_hugepages_node(*args):
+            taskset_param = ' -c'
+
+        # prepend taskset only if valid coremask was specified in testbed
+        with settings(host_string=host_string):
```
Why is host_string=host_string needed?

semihalf-zmuda-wojciech replied:
The settings() context manager temporarily overrides env variable passed as argument. In this case we override env.host_string (the left hand operand) with host_string (the right hand operand), which is iterated through args in the for loop (line 817). I agree that a statement like host_string=host_string seems a little weird, but this is convention used throughout the file (lines 29, 624, 737, 776) , so I mimicked the style.

More info on settings() and host_string:
http://docs.fabfile.org/en/latest/usage/env.html#host-string
http://docs.fabfile.org/en/latest/api/core/context_managers.html

srajag commented on fabfile/tasks/helpers.py:
```diff
@@ -806,6 +806,38 @@ def setup_hugepages_node(*args):
+        # prepend taskset only if valid coremask was specified in testbed
+        with settings(host_string=host_string):
+            sudo('sed -i \'s/command=/command=taskset%s %s /\' %s' \
+                %(taskset_param, coremask, vrouter_file))
```
What happens if the core mask is not well formed - will we end up using the default core mask? Instead, can we make the provisioning fail? Also, can the core mask be specified in the same place as the huge pages percentage specification? I don;t think we should require a separate fab command for this.

semihalf-zmuda-wojciech replied:
> What happens if the core mask is not well formed

Actually there is not thorough validation, so coremask like "asdf" is accepted and taskset will fail. I assumed this option will be used by someone who knows what they're doing. Would you like to get proper validation?

> can we make the provisioning fail?

Yes, it is possible. Would you like make coremask a mandatory option? Currently, when no coremask is specified (or it is empty), it is simply ignored and taskset does not get prepended to the command line in the .ini file. It is totally harmless to the setup, as vRouter will take care of starting without coremask provided (4 cores by default). Would you like to change that behavior?

> can the core mask be specified in the same place as the huge pages

Could you be more specific? It is in the same script, in the separate function, just below the huge page function.

> I don;t think we should require a separate fab command for this.

There is no separate command. It all gets executed during fab setup_all. If I do not understand you correctly, will you please elaborate more on your requirements?

srajag replied:
I think we can validate core mask by doing a taskset with the mask specified (and some dummy command) and verifying whether it succeeded. If it didn't, it would be better to fail the provisioning. 

If the core mask is specifed in the same line as huge pages (per server), it is fine.

semihalf-zmuda-wojciech replied:
> If the core mask is specifed in the same line as huge pages (per server), it is fine.

Yes, the section in testbed.py looks like this:
env.dpdk = {
    host1: { 'huge_pages': '50', 'coremask': '0,2-6' },
    host2: { 'huge_pages': '50', 'coremask': '0xf' },
}

I'm testing changes you asked for (validation and making the parameter mandatory) and will commit it soon.

semihalf-zmuda-wojciech committed:
Validate affinity mask specified in testbed.

Validation is done with regular expressions. If mask is not provied
as a hexadecimal number (eg. 0xFF, 0xab) or as a list of CPUs (eg. 0,2-6)
the script exits.

semihalf-zmuda-wojciech committed:
Validate core mask by calling taskset and checking exit code.

Stop provisioning if empty or invalid value has been provided in testbed.

semihalf-zmuda-wojciech committed:
Add core mask to example testbeds.
